src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -385,45 +388,81 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
   abstract class GlobalPhase(prev: Phase) extends Phase(prev) {
     phaseWithId(id) = this
 
-    def run(): Unit = {
-      echoPhaseSummary(this)
-      currentRun.units foreach applyPhase
-    }
-
     def apply(unit: CompilationUnit): Unit
 
     /** Is current phase cancelled on this unit? */
     def cancelled(unit: CompilationUnit) = {
+      assertOnMain()
       // run the typer only if in `createJavadoc` mode
       val maxJavaPhase = if (createJavadoc) currentRun.typerPhase.id else currentRun.namerPhase.id
       reporter.cancelled || unit.isJava && this.id > maxJavaPhase
     }
 
-    final def withCurrentUnit(unit: CompilationUnit)(task: => Unit): Unit = {
-      if ((unit ne null) && unit.exists)
-        lastSeenSourceFile = unit.source
+    // Method added to allow stacking functionality on top of the`run` (e..g measuring run time).
+    // Overriding `run` is now not allowed since we want to be in charge of how units are processed.
+    def wrapRun(code: => Unit): Unit = code
+
+    def afterUnit(unit: CompilationUnit): Unit = {}
+
+    final def run(): Unit = wrapRun {
+      assertOnMain()
+
+      if (isDebugPrintEnabled) inform("[running phase " + name + " on " + currentRun.size +  " compilation units]")
+
+      implicit val ec: ExecutionContextExecutorService = createExecutionContext()
+
+      def task(unit: CompilationUnit): Reporter = {
+        processUnit(unit)
+        afterUnit(unit)
+        reporter
+      }
+
+      val futures = currentRun.units.collect {
+        case unit if !cancelled(unit) =>
+          if(isParallel) Future(task(unit)) else Future.fromTry(scala.util.Try(asWorkerThread(task(unit))))
+      }
 
-      if (settings.debug && (settings.verbose || currentRun.size < 5))
-        inform("[running phase " + name + " on " + unit + "]")
-      if (!cancelled(unit)) {
-        currentRun.informUnitStarting(this, unit)
-        try withCurrentUnitNoLog(unit)(task)
-        finally currentRun.advanceUnit()
+      futures.foreach { future =>
+        val workerReporter = Await.result(future, Duration.Inf)
+        workerReporter.asInstanceOf[BufferedReporter].flushTo(reporter)
       }
     }
 
-    final def withCurrentUnitNoLog(unit: CompilationUnit)(task: => Unit): Unit = {
+    final def applyPhase(unit: CompilationUnit): Unit = {
+      assertOnWorker()
+      if (!cancelled(unit)) processUnit(unit)
+    }
+
+    private def processUnit(unit: CompilationUnit): Unit = {
+      assertOnWorker()
+
+      reporter = new BufferedReporter
+
+      if (isDebugPrintEnabled) inform("[running phase " + name + " on " + unit + "]")
+
       val unit0 = currentUnit
+
       try {
+        if ((unit ne null) && unit.exists) lastSeenSourceFile = unit.source
         currentRun.currentUnit = unit
```

Comment:
this needs to be parallel?

Reply:
Unluckily yes, at least as long as we have `currentUnit`.
`lastSeenSourceFile` is used inside error reporting which happens inside worker thread.
If we would make it global it could report errors on incorrect files.

That said I do not understand why we need it in first place.
Comment in Global.scala says:

```scala
  /** There are common error conditions where when the exception hits
   *  here, currentRun.currentUnit is null.  This robs us of the knowledge
   *  of what file was being compiled when it broke.  Since I really
   *  really want to know, this hack.
   */
  protected var lastSeenSourceFile: SourceFile = NoSourceFile
```

But i cannot see how `lastSeenSourceFile` can be set when `currentRun.currentUnit` is null.